dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu_test.cu

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <vector>
+#include <tuple>
+#include "dali/kernels/scratch.h"
+#include "dali/kernels/tensor_shape.h"
+#include "dali/kernels/common/copy.h"
+#include "dali/kernels/test/tensor_test_utils.h"
+#include "dali/kernels/test/kernel_test_utils.h"
+#include "dali/kernels/imgproc/color_manipulation/brightness_contrast_gpu.h"
+
+namespace dali {
+namespace kernels {
+namespace brightness_contrast {
+namespace test {
+
+namespace {
+
+constexpr size_t kNdims = 3;
+
+/**
+ * Rounding to nearest even (like GPU does it)
+ */
+template <class In, class Out>
+std::enable_if_t<std::is_integral<Out>::value, Out> custom_round(float val) {
+  return static_cast<Out>(std::nearbyint(val));
+}
+
+
+template <class In, class Out>
+std::enable_if_t<!std::is_integral<Out>::value, Out> custom_round(float val) {
+  return val;
+}
+
+
+}  // namespace
+
+template <class InputOutputTypes>
+class BrightnessContrastGpuTest : public ::testing::Test {
+  using In = typename InputOutputTypes::In;
+  using Out = typename InputOutputTypes::Out;
+
+ public:
+  BrightnessContrastGpuTest() {
+    input_host_.resize(dataset_size());
+  }
+
+
+  void SetUp() final {
+    std::mt19937_64 rng;
+    UniformRandomFill(input_host_, rng, 0., 10.);
+    calc_output(0);
+    CUDA_CALL(cudaMalloc(&input_device_, sizeof(In) * dataset_size()));
+    CUDA_CALL(cudaMemcpy(input_device_, input_host_.data(), input_host_.size() * sizeof(In),
+                         cudaMemcpyDefault));
+    CUDA_CALL(cudaMallocManaged(&output_, dataset_size() * sizeof(Out)));
+    cudaDeviceSynchronize();
+
+    verify_test();
+  }
+
+
+  In *input_device_;
+  Out *output_;
+  std::vector<In> input_host_;
+  std::vector<Out> ref_output_;
+  std::vector<TensorShape<kNdims>> shapes_ = {{2, 4, 3}};
+  std::vector<float> brightness_ = {4};
+  std::vector<float> contrast_ = {3};
+
+
+  void verify_test() {
+    assert(shapes_.size() == brightness_.size());
```

Comment:
use ASSERT_EQ from gtest

Reply:
That's only debug info for me, if this fires, that means that something is broken in implementation of the test, not the tested code itself. I like to separate these two...

Reply:
I normally assume:
- ASSERT_XX: Assumptions that must always satisfy, otherwise the test should be aborted right away
- EXPECT_XX: Those are test expectation, the test could continue but it should be marked as failed

Reply:
That's correct. Thing is, that if any of these asserts fail, the test implementation itself has no sense and it wouldn't test anything. It's not testing anything, it's verifying, if the test is implemented with proper asserts

Reply:
so, that's what I normally assert with ASSERT_XXX